core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -43,19 +45,179 @@
+                and 'question_skill_links' to a list of QuestionSkillLinkModel.
+
+        Returns:
+            Result[opportunity_models.SkillOpportunityModel, Exception].
+            Result object that contains SkillOpportunityModel when the operation
+            is successful and Exception when an exception occurs.
+        """
+        try:
+            skill_opportunity = opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                skill_id=skills_and_links['skill'].id,
+                skill_description=skills_and_links['skill'].description,
+                question_count=GenerateSkillOpportunityModelJob
+                    ._count_unique_question_ids(
+                        skills_and_links['question_skill_links']
+            ))
+            skill_opportunity.validate() # type: ignore[no-untyped-call]
+            skill_opportunity_model = opportunity_models.SkillOpportunityModel(
+                id=skill_opportunity.id,
+                skill_description=skill_opportunity.skill_description,
+                question_count=skill_opportunity.question_count
+            )
+            skill_opportunity_model.update_timestamps()
+            return result.Ok(skill_opportunity_model)
+        except Exception as e:
+            return result.Err(e)
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of 'SUCCESS' or 'FAILURE' results from
+        generating SkillOpportunityModel.
+
+        Returns:
+            PCollection. A PCollection of 'SUCCESS' or 'FAILURE' results from
+            generating SkillOpportunityModel.
+        """
+        question_skill_link_models = (
+            self.pipeline
+            | 'Get all non-deleted QuestionSkillLinkModels' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSkillLinkModel.get_all(
+                        include_deleted=False))
+                )
+            | 'Group QuestionSkillLinkModels by skill id' >>
+                beam.GroupBy(lambda n: n.skill_id)
+        )
+
+        skills = (
+            self.pipeline
+            | 'Get all non-deleted SkillModels' >> (
+                ndb_io.GetModels(
+                    skill_models.SkillModel.get_all(include_deleted=False)))
+            | 'Get skill from model' >> beam.Map(
+                skill_fetchers.get_skill_from_model)
+            | 'Group Skills by skill id' >> beam.GroupBy(lambda m: m.id)
+        )
+
+        skills_with_question_counts = (
+            {
+                'skill': skills,
+                'question_skill_links': question_skill_link_models
+            }
+            | 'Merge by skill id' >> beam.CoGroupByKey()
+            | beam.Map(
+                lambda n: {
+                    'skill': n[1]['skill'][0][0],
+                    'question_skill_links': list(itertools.chain.from_iterable(
+                        n[1]['question_skill_links']))
+                }
+            )
```

Comment:
```suggestion
            | 'Merge by skill ID' >> beam.CoGroupByKey()
            | 'Remove skill IDs' >> beam.Values()
            | beam.Map(
                lambda object: {
                    'skill': object['skill'][0][0],
                    'question_skill_links': list(itertools.chain.from_iterable(
                        object['question_skill_links']))
                }
            )
```

Reply:
Done

Reply:
I think it improves the readabiliy, you can add an ignore and comment explaining it 
![image](https://user-images.githubusercontent.com/10142938/146930124-68c1eef6-322b-4917-aea0-6da267559334.png)

Reply:
Ok, done! Thanks